pkg/virt-manifest/rest/manifest.go

```diff
@@ -0,0 +1,106 @@
+package rest
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/emicklei/go-restful"
+	ghodss_yaml "github.com/ghodss/yaml"
+	"k8s.io/client-go/pkg/util/yaml"
+
+	"kubevirt.io/kubevirt/pkg/api/v1"
+	"kubevirt.io/kubevirt/pkg/logging"
+	"kubevirt.io/kubevirt/pkg/virt-handler/virtwrap"
+	"kubevirt.io/kubevirt/pkg/virt-manifest"
+)
+
+const (
+	MIME_YAML = "application/yaml"
+)
+
+type manifest struct {
+	libvirtUri    string
+	libvirtUser   string
+	libvirtPasswd string
+}
+
+type entityYAMLAccess struct {
+	ContentType string
+}
+
+// Read isn't currently used in this implementation
+func (e entityYAMLAccess) Read(req *restful.Request, v interface{}) error {
+	return fmt.Errorf("not implemented")
+}
+
+// Write marshals the value to YAML
+func (e entityYAMLAccess) Write(resp *restful.Response, status int, v interface{}) error {
+	data, err := ghodss_yaml.Marshal(v)
+	if err != nil {
+		return err
+	}
+	resp.WriteHeader(status)
+	resp.Write(data)
+	return nil
+}
+
+func ManifestService(libvirtUri string, libvirtUser string, libvirtPasswd string) (*restful.WebService, error) {
+	ws := new(restful.WebService)
+	m := manifest{
+		libvirtUri:    libvirtUri,
+		libvirtUser:   libvirtUser,
+		libvirtPasswd: libvirtPasswd}
+	restful.RegisterEntityAccessor(MIME_YAML, entityYAMLAccess{})
+	ws.Path("/").
+		Consumes(restful.MIME_JSON, MIME_YAML).
+		Produces(restful.MIME_JSON, MIME_YAML)
+	ws.Route(ws.GET("/api/v1/status").To(m.alive))
+	ws.Route(ws.POST("/api/v1/manifest").To(m.mapManifest))
+	return ws, nil
+}
+
+func (m manifest) alive(request *restful.Request, response *restful.Response) {
+	res := map[string]interface{}{"status": "ok"}
+	response.WriteEntity(res)
+}
+
+func (m manifest) mapManifest(request *restful.Request, response *restful.Response) {
```

Comment:
you can remove the api logic from this call by using our rest.endpoints code. The result would look something like this:


```golang
// You can for instance return a middleware.NewBadRequestError(err), to return 400
func mapManifest(_ context.Context, request interface{}) (interface{}, error) {

	log := logging.DefaultLogger()

	vm := request.(*PutObject).Payload.(*v1.VM)

	if vm.Spec.Domain == nil {
		vm.Spec.Domain = new(v1.DomainSpec)
	}

	err = virt_manifest.AddMinimalDomainSpec(vm.Spec.Domain)
	if err != nil {
		log.Error().Reason(err).Msg("Unable to map domain")
		return nil, middleware.NewInternalServerError(err)
	}

	connection, err := virtwrap.NewConnection(m.libvirtUri, m.libvirtUser, m.libvirtPasswd, 0)
	if err != nil {
		return nil, middleware.NewInternalServerError(err)
	}

	mappedVm, err := virt_manifest.MapVM(connection, vm)
	if err != nil {
		return nil ,middleware.NewInternalServerError(err)
	} 
       return mappedVM, nil
}

ctx := context.Background()

// This include serialization/deserialization and validation of v1.VM from json/yaml to json/yaml
target := MakeGoRestfulWrapper(NewHandlerBuilder().Post(&v1.VM{}).Endpoint(mapManifest).Build(ctx))


// Wire go-restful like you did in the main
ws.Path("/").
	Consumes(restful.MIME_JSON, MIME_YAML).
		Produces(restful.MIME_JSON, MIME_YAML)
	ws.Route(ws.GET("/api/v1/status").To(m.alive))


ws.Route(ws.POST("/api/v1/manifest").To(target))
```